Content.Server/Ninja/Systems/NinjaGlovesSystem.cs

```diff
@@ -0,0 +1,138 @@
+using Content.Shared.DoAfter;
+using Content.Shared.Interaction.Components;
+using Content.Shared.Interaction.Events;
+using Content.Shared.Ninja.Components;
+using Content.Shared.Ninja.Systems;
+using Content.Shared.Popups;
+using Content.Shared.Research.Components;
+using Content.Shared.Toggleable;
+
+namespace Content.Server.Ninja.Systems;
+
+public sealed class NinjaGlovesSystem : SharedNinjaGlovesSystem
+{
+    [Dependency] private readonly EmagProviderSystem _emagProvider = default!;
+    [Dependency] private readonly SharedBatteryDrainerSystem _drainer = default!;
+    [Dependency] private readonly SharedStunProviderSystem _stunProvider = default!;
+    [Dependency] private readonly SpaceNinjaSystem _ninja = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<NinjaGlovesComponent, ToggleActionEvent>(OnToggleAction);
+    }
+
+    /// <summary>
+    /// Toggle gloves, if the user is a ninja wearing a ninja suit.
+    /// </summary>
+    private void OnToggleAction(EntityUid uid, NinjaGlovesComponent comp, ToggleActionEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        args.Handled = true;
+
+        var user = args.Performer;
+        // need to wear suit to enable gloves
+        if (!TryComp<SpaceNinjaComponent>(user, out var ninja)
+            || ninja.Suit == null
+            || !HasComp<NinjaSuitComponent>(ninja.Suit.Value))
+        {
+            Popup.PopupEntity(Loc.GetString("ninja-gloves-not-wearing-suit"), user, user);
+            return;
+        }
+
+        var enabling = comp.User == null;
+        Appearance.SetData(uid, ToggleVisuals.Toggled, enabling);
+        var message = Loc.GetString(enabling ? "ninja-gloves-on" : "ninja-gloves-off");
+        Popup.PopupEntity(message, user, user);
+
+        if (enabling)
+        {
+            comp.User = user;
+            _ninja.AssignGloves(ninja, uid);
+            // set up interaction relay for handling glove abilities, comp.User is used to see the actual user of the events
+            // TODO: remove, bad
+            Interaction.SetRelay(user, uid, EnsureComp<InteractionRelayComponent>(user));
+            Dirty(comp);
+
+            var drainer = EnsureComp<BatteryDrainerComponent>(user);
+            var stun = EnsureComp<StunProviderComponent>(user);
+            if (_ninja.GetNinjaBattery(user, out var battery, out var _))
+            {
+                _drainer.SetBattery(drainer, battery);
+                _stunProvider.SetBattery(stun, battery);
+            }
+
+            var emag = EnsureComp<EmagProviderComponent>(user);
+            _emagProvider.SetWhitelist(user, comp.DoorjackWhitelist, emag);
+
+        }
+        else
+        {
+            DisableGloves(uid, comp);
+        }
+    }
+
+    /// <inheritdoc/>
+    protected override void OnDownloadDoAfter(EntityUid uid, NinjaDownloadComponent comp, DownloadDoAfterEvent args)
```

Comment:
im gonna assume the subscription is in shared, in which case please move it to here, the event should be subscribed in the same assembly it is handled (i dont want to do this for all of em, so just check the rest of your files)